matRad_generateStf.m

```diff
@@ -313,36 +304,56 @@
                 
             else % target not hit
-                stf(i).ray(j) = [];
-            end
-                        
+                stf(i).ray(j)       = [];
+           end
+                 
         end
 
         % book keeping
         stf(i).numOfRays = size(stf(i).ray,2);
         for j = 1:stf(i).numOfRays
+            
             stf(i).numOfBixelsPerRay(j) = numel(stf(i).ray(j).energy);
-            for k = 1:stf(i).numOfBixelsPerRay(j)
-                stf(i).ray(j).focusIx(k) = 
+            focusIx  =  machine.meta.defaultFociIndex * ones(stf(i).numOfBixelsPerRay(j),1);
+            [~, energyIx] = min(abs(bsxfun(@minus,[machine.data.energy]',...
+                            repmat(stf(i).ray(j).energy,length([machine.data]),1))));
+                   
+            % get for each spot the focus index
+            Cnt = 1; 
+            for k = energyIx
+                for currFoci =  machine.meta.defaultFociIndex:1:size(machine.data(k).initFocus,2)
+            
+                      SigmaIniAtIsoCenter = interp1(machine.data(k).initFocus(currFoci).dist,...
+                                                    machine.data(k).initFocus(currFoci).sigma,...
+                                                    machine.meta.BAMStoIsoDist);
+                           
+                     if SigmaIniAtIsoCenter > machine.meta.minIniBeamSigma 
+                            focusIx(Cnt)= currFoci;
+                        break;
+                     end
+                end
+                 Cnt = Cnt + 1;
             end
+            
+            stf(i).ray(j).focusIx = focusIx';
+            
         end
 
     elseif strcmp(stf(i).radiationMode,'photons')
         % set dummy values for photons
         for j = 1:stf(i).numOfRays
             
-            % ray tracing necessary to determine SSD
-            [alphasSSD,~,rhoSSD,~] = matRad_siddonRayTracer(pln.isoCenter, ...
-                                        ct.resolution, ...
-                                        stf(i).sourcePoint, ...
-                                        stf(i).ray(j).targetPoint, ...
-                                        {voiSSD});
-
-            ixSSD = find(rhoSSD{1} > 0,1,'first');
-            if ixSSD == 1
-                warning('Surface for SSD calculation starts directly in first voxel of CT\n');
-            end
-            stf(i).SSD(j) = stf(i).SAD * alphasSSD(ixSSD);
+             % ray tracing necessary to determine depth of the target
+            [~,~,rho,~,ix] = matRad_siddonRayTracer(stf(i).isoCenter, ...
+                                 ct.resolution, ...
+                                 stf(i).sourcePoint, ...
+                                 stf(i).ray(j).targetPoint, ...
+                                 {ct.cube,voiTarget});
+            
+ 
+            DensityThreshold = 0.05;
+            ixSSD = find(rho{1} > DensityThreshold,1,'first');
+            stf(i).ixSSD(j) = ix(ixSSD);
             
```

Comment:
investigate which part of the code can be combined for photons and particles. we are doing a lot of stuff twice which complicates code maintenance...